### PR TITLE
FFDC: Collect current in-memory sessions for BMC dump

### PR DESCRIPTION
**Pull request by raviteja-b**
Currently bmcweb_persistent_data.json file collected as part of BMC
dump, but this file does not contain all existing redfish sessions data.

This commit is to collect all sessions into bmcweb_current_session_persistent_data.json
during bmc dump creation, on receiving SIGUSR1 from dump manager.

Dump manager sends SIGUSR to bmcweb while creating bmc dump,
bmcweb catches this signal and dumps sessions data into a file.
This sessions data will be useful for debugging redfish session related issues.
Added event config and event subscriptions data along with session data

Tested by:

Kill  -10 bmcweb-pid
Check content of bmcweb_current_session_persistent_data.json

### CODE REVIEW

**Comment by gtmills** [comment]
@raviteja-b Why do we need this?

**Comment by raviteja-b** [comment]
> @raviteja-b Why do we need this?

this is need to collect existing sessions during bmc dump collection, needed for debugging issue

**Review by sunharis — APPROVED** [review]

**Review by asmithakarun — APPROVED** [review]

**Comment by gtmills** [comment]
My concerns are: 
The introduction of something that could cause bmcweb to core dump in the middle of running 
performance 
another patch we carry downstream

File: http/http_server.hpp [hunk 1]
```diff
@@ -163,6 +169,16 @@ class Server
                     }
                     this->startAsyncWaitForSignal();
                 }
+#ifdef BMCWEB_ENABLE_IBM_MANAGEMENT_CONSOLE
+                if (signalNo == SIGUSR1)
```
Comment (gtmills) [inline http/http_server.hpp:173]: We can't kill bmcweb to get this data

**Review by gtmills — CHANGES_REQUESTED** [review]

**Comment by gtmills** [comment]
Planning on discussing this at the SoS overflow on 04/04/2022. I feel this violates principle of least astonishment (using kill -10 to get data) and getting this data via a D-Bus signal or a property or method fits much better with the OpenBMC architecture. Could we ask in Discord?

**Comment by amboar** [comment]
> Planning on discussing this at the SoS overflow on 04/04/2022. I feel this violates principle of least astonishment (using kill -10 to get data) and getting this data via a D-Bus signal or a property or method fits much better with the OpenBMC architecture. Could we ask in Discord?

I think it depends on *why* and *when* we want the data. Are we trying to acquire it because bmcweb is unresponsive? If so, it might not be a good idea to do it via D-Bus, as that may be unreliable. I haven't looked at the code, but the signal delivery might also be unreliable if the code uses signalfd() as we may be stuck somewhere else and not poll()ing the descriptor.

Have we considered just taking a core dump?

File: http/http_server.hpp (hunk 1 above)
Comment (rsnair) [reply http/http_server.hpp:173]: SIGUSR1 is a user signal. This wouldn't kill any process. Its commonly used by linux applications to override and implement specific actions.

**Comment by rsnair** [comment]
@amboar 

> Are we trying to acquire it because bmcweb is unresponsive?  

No - The commit is just catching a SIGUSR1 to dump a specific in-memory  application data to  a file. The only use case for this is while collecting a user initiated BMCDUMP.

File: include/persistent_data.hpp [hunk 2]
```diff
@@ -190,6 +192,60 @@ class ConfigFile
...
+        nlohmann::json& sessions = data["sessions"];
+        sessions = nlohmann::json::array();
+        for (const auto& p : SessionStore::getInstance().authTokens)
+        {
+            if (p.second->persistence !=
+                persistent_data::PersistenceType::SINGLE_REQUEST)
+            {
+                sessions.push_back({
+                    {"unique_id", p.second->uniqueId},
+                    {"session_token", p.second->sessionToken},
```
Comment (sunharis) [inline include/persistent_data.hpp:237]: Please remove session_token and csrf_token

File: include/persistent_data.hpp [hunk 3]
```diff
@@ -27,6 +27,8 @@ class ConfigFile
   public:
     // todo(ed) should read this from a fixed location somewhere, not CWD
     static constexpr const char* filename = "bmcweb_persistent_data.json";
+    static constexpr const char* dumpFilename =
+        "bmcweb_current_session_persistent_data.json";
```
Comment (joseph-reynolds) [inline include/persistent_data.hpp:31]: The filename is misleading.  Perhaps: bmcweb_current_session_snapshot.json.
Maybe also include the BMC's current TOD in the file so we know when the snapshot was taken?

File: include/persistent_data.hpp (hunk 2 above)
Comment (joseph-reynolds) [inline include/persistent_data.hpp:237]: We discussed in today's SoS to exclude this information, and only provide the same information available under /redfish/v1/SessionService/Sessions/**.

File: include/persistent_data.hpp [hunk 4]
```diff
@@ -190,6 +192,60 @@ class ConfigFile
...
+        for (const auto& p : SessionStore::getInstance().authTokens)
+        {
+            if (p.second->persistence !=
+                persistent_data::PersistenceType::SINGLE_REQUEST)
+            {
+                sessions.push_back({
+                    {"unique_id", p.second->uniqueId},
+                    {"session_token", p.second->sessionToken},
+                    {"username", p.second->username},
+                    {"csrf_token", p.second->csrfToken},
```
Comment (joseph-reynolds) [inline include/persistent_data.hpp:239]: Should we also exclude this?

File: include/persistent_data.hpp [hunk 5]
```diff
@@ -190,6 +192,60 @@ class ConfigFile
...
+                    {"client_ip", p.second->clientIp},
+                    {"client_id", p.second->clientId},
+                });
+            }
+        }
+        persistentFile << data;
+    }
```
Comment (joseph-reynolds) [inline include/persistent_data.hpp:246]: I am curious what happens if there is a problem writing the dump file.  (It looks like this code was just copied from the writeData() function below.)  If the file cannot be written, is an error indicated?

**Review by joseph-reynolds — COMMENTED** [review]
Looks good to me, pending the change to not write the session token and csrf.  But I am not familiar with BMCWeb signal handling.

File: http/http_server.hpp (hunk 1 above)
Comment (joseph-reynolds) [reply http/http_server.hpp:173]: As mentioned in today's SoS, the Linux `kill` command sends a signal to the process (https://en.wikipedia.org/wiki/Signal_(IPC)).  It does not necessarily send the terminate or kill signal.  And it is common to make long-running daemons respond to various signals

File: include/persistent_data.hpp (hunk 2 above)
Comment (raviteja-b) [reply include/persistent_data.hpp]: done

File: include/persistent_data.hpp (hunk 3 above)
Comment (raviteja-b) [reply include/persistent_data.hpp:31]: yes this name bmcweb_current_session_snapshot.json makes sense

File: include/persistent_data.hpp (hunk 4 above)
Comment (raviteja-b) [reply include/persistent_data.hpp]: yes

File: include/persistent_data.hpp [hunk 6]
```diff
@@ -190,6 +192,60 @@ class ConfigFile
...
+
+        nlohmann::json& sessions = data["sessions"];
+        sessions = nlohmann::json::array();
+        for (const auto& p : SessionStore::getInstance().authTokens)
+        {
+            if (p.second->persistence !=
+                persistent_data::PersistenceType::SINGLE_REQUEST)
+            {
+                sessions.push_back({
+                    {"unique_id", p.second->uniqueId},
```
[the remark this answers is not in the document]
Comment (raviteja-b) [reply include/persistent_data.hpp:218]: ok

File: include/persistent_data.hpp (hunk 5 above)
Comment (raviteja-b) [reply include/persistent_data.hpp:226]: I see any  need to log error in this case of failure

File: include/persistent_data.hpp (hunk 3 above)
Comment (gtmills) [reply include/persistent_data.hpp:31]: > bmcweb_current_session_snapshot.json 

+1

File: include/persistent_data.hpp (hunk 5 above)
Comment (gtmills) [reply include/persistent_data.hpp:226]: If it this important of information seems like we would want a log knowing it failed to write

File: include/persistent_data.hpp (hunk 3 above)
Comment (raviteja-b) [reply include/persistent_data.hpp]: done

**Comment by raviteja-b** [comment]
CI failing for redfish-core/lib/assembly.hpp file, which is not related this PR.
Does bmcweb CI broken?

**Comment by raviteja-b** [comment]
@gtmills how to fix this CI issue?
seems this PR https://github.com/ibm-openbmc/bmcweb/pull/224 causing CI failure?

**Comment by aahmed-2** [comment]
jenkins run tests please

**Comment by raviteja-b** [comment]
jenkins please run hw ci

**Review by sunharis — APPROVED** [review]
Seems good to go after CI passes

**Comment by aahmed-2** [comment]
jenkins run test please

**Comment by raviteja-b** [comment]
jenkins run test please

**Comment by raviteja-b** [comment]
jenkins run test please

**Comment by gtmills** [comment]
Although I still am struggling to see the need for this, the team supporting HMC <--> feels strongly this is needed. Approved.

**Comment by gtmills** [comment]
Please continue discussing upstream.

**Review by gtmills — APPROVED** [review]

**Comment by gtmills** [comment]
@rfrandse can you squash this when you merge to get rid of that 2nd empty commit: 
https://github.com/ibm-openbmc/bmcweb/pull/310/commits

**Comment by raviteja-b** [comment]
I have added new  commit to this PR to add  event subscription data as well.

**Pull request merged**